src/util/log.c

```diff
@@ -0,0 +1,648 @@
+        }
+
+        log->mem_fd = mem_fd;
+        log->map = p;
+        mem_fd = -1;
+        return true;
+}
+
+static int log_fd_send(int destination_fd, int payload_fd) {
+        union {
+                char buffer[CMSG_SPACE(sizeof(int))];
+                struct cmsghdr cmsg;
+        } control;
+        struct msghdr msg;
+        ssize_t l;
+
+        control.cmsg.cmsg_len = CMSG_LEN(sizeof(int));
+        control.cmsg.cmsg_level = SOL_SOCKET;
+        control.cmsg.cmsg_type = SCM_RIGHTS;
+        memcpy(CMSG_DATA(&control.cmsg), &payload_fd, sizeof(int));
+
+        msg = (struct msghdr){
+                .msg_control = &control.cmsg,
+                .msg_controllen = control.cmsg.cmsg_len,
+        };
+
+        l = sendmsg(destination_fd, &msg, MSG_NOSIGNAL);
+        if (l)
+                return (l < 0) ? error_origin(-errno) : error_origin(-ENOTRECOVERABLE);
+
+        return 0;
+}
+
+static int log_loop_send(int destination_fd, const void *blob, size_t n_blob) {
+        ssize_t l;
+
+        while (n_blob > 0) {
+                l = send(destination_fd, blob, n_blob, MSG_NOSIGNAL);
+                if (l < 0)
+                        return error_origin(-errno);
+                else if (l == 0 || l > (ssize_t)n_blob)
+                        return error_origin(-ENOTRECOVERABLE);
+
+                blob += l;
+                n_blob -= l;
+        }
+
+        return 0;
+}
+
+static int log_stream_send(Log *log) {
+        bool log_warn = false;
+        const void *blob = log->map;
+        size_t n_blob = log->offset;
+        ssize_t l;
+        int r, v;
+
+        /*
+         * Stream sockets are a bit nasty since they lack atomic writes. Hence,
+         * whenever we end up with a short-write in lossy mode, we need to
+         * finish the current message in synchronous mode, and then log a
+         * warning. Otherwise, we end up with a garbled output.
+         *
+         * Now, we remember whenever we dropped message. From then on, we only
+         * ever write messages if the output buffer of our socket is empty.
+         * This guarantees the operation will be non-blocking, as long as
+         * suitably sized log messages are used.
+         */
+
+        if (log->lossy) {
+                if (log->n_dropped) {
+                        r = ioctl(log->log_fd, SIOCOUTQ, &v);
+                        if (r < 0) {
+                                return error_origin(-errno);
+                        } else if (v) {
+                                ++log->n_dropped;
+                                return 0;
+                        }
+                }
+
+                l = send(log->log_fd, blob, n_blob, MSG_NOSIGNAL | MSG_DONTWAIT);
+                if (l >= (ssize_t)n_blob) {
+                        return 0;
+                } else if (l < 0) {
+                        if (errno != EAGAIN)
+                                return error_origin(-errno);
+
+                        l = 0;
```

Comment:
So in case of a short write, we resubmt the prefix that was already pushed to the socket?

Reply:
Errr, what prefix was already pushed?

Reply:
Ignore that, misread.